File: simulation.py
"""Phase 3: Tick-based simulation with LLM-powered agent brains."""

from __future__ import annotations
import json
import random
import asyncio
from models import (
    AgentProfile, CampaignSeed, Interaction, SimulationResult,
    ActionType, Mood,
)
from social_graph import SocialGraph
from prompts import agent_brain, build_context
from event_bus import EventBus
from llm import chat


async def run_simulation(
    seed: CampaignSeed,
    agents: list[AgentProfile],
    graph: SocialGraph,
    num_ticks: int,
    generation: int,
    bus: EventBus,
) -> SimulationResult:
    interactions: list[Interaction] = []
    agent_map = {a.id: a for a in agents}
    seen: set[str] = set()

    for tick in range(num_ticks):
        await bus.emit("tick", {"tick": tick, "total": num_ticks})

        if tick == 0:
            influencers = {a.id for a in graph.get_influencers(5)}
            sample_size = max(1, len(agents) * 30 // 100)
            random_set = {a.id for a in random.sample(agents, min(sample_size, len(agents)))}
            active_ids = list(influencers | random_set)
        else:
            spreaders = {ix.agent_id for ix in interactions
                         if ix.action in (ActionType.SHARE, ActionType.QUOTE_SHARE, ActionType.COMMENT)}
            active_ids = []
            for sid in spreaders:
                for fid in graph.get_followers_of(sid):
                    if fid not in seen:
                        active_ids.append(fid)
            active_ids = list(set(active_ids))

        if not active_ids:
            continue

        random.shuffle(active_ids)

        # Process agents in concurrent batches of 5
        batch_size = 5
        for i in range(0, len(active_ids), batch_size):
            batch = active_ids[i:i + batch_size]
            tasks = []
            for aid in batch:
                if aid in seen:
                    continue
                seen.add(aid)
                agent = agent_map.get(aid)
                if not agent:
                    continue
                visible = [ix for ix in interactions if ix.action != ActionType.IGNORE][-10:]
                context = build_context(seed, visible, agent_map)
                tasks.append(process_agent(agent, context, tick, bus))

            results = await asyncio.gather(*tasks, return_exceptions=True)
            for r in results:
                if isinstance(r, Interaction):
                    interactions.append(r)
                elif isinstance(r, Exception):
                    print(f"  [WARN] Agent failed: {r}")

    return SimulationResult(seed=seed, generation=generation, interactions=interactions)


async def process_agent(
    agent: AgentProfile,
    context: str,
    tick: int,
    bus: EventBus,
) -> Interaction:
    prompt = agent_brain(agent, context)

    raw = await chat(
        system="You are a social-media user simulator. Return ONLY valid JSON. No markdown.",
        user=prompt,
        max_tokens=512,
    )

    cleaned = raw.replace("```json", "").replace("```", "").strip()
    data = json.loads(cleaned)

    action = ActionType(data.get("action", "ignore"))
    new_mood = next((m for m in Mood if m.value == data.get("new_mood", "neutral")), Mood.NEUTRAL)
    agent.mood = new_mood

    interaction = Interaction(
        agent_id=agent.id,
        tick=tick,
        action=action,
        content=data.get("content", ""),
        reasoning=data.get("reasoning", ""),
        new_mood=new_mood,
    )

    await bus.emit("reaction", {
        "agentId": agent.id,
        "agentName": agent.name,
        "action": action.value,
        "content": interaction.content,
        "reasoning": interaction.reasoning,
        "mood": new_mood.value,
        "tick": tick,
    })

    return interaction

File: src/main/kotlin/com/resonance/Simulation.kt
package com.resonance

import ai.koog.agents.core.agent.AIAgent
import ai.koog.prompt.executor.LLMPromptExecutor
import ai.koog.prompt.llm.OpenAIModels
import kotlinx.serialization.json.Json
import kotlin.random.Random

// ==========================================================================
// Phase 3: Tick-based Simulation Engine
//
// Each synthetic persona gets its own Koog AIAgent call to decide how they
// react to the post. The simulation runs in ticks:
//
//   Tick 0: Seed viewers (influencers + random sample) see the post directly.
//   Tick 1+: Agents see the post ONLY if someone they follow shared/commented.
//            This is where second-order effects + viral cascades emerge.
//
// Key emergent behaviors:
//   - Pile-ons: if influencer mocks, followers join in
//   - Viral spread: shares propagate to new subgraphs
//   - Message distortion: quote_shares reframe the original content
//   - Echo chambers: agents in tight clusters amplify each other
// ==========================================================================

suspend fun runSimulation(
    executor: LLMPromptExecutor,
    seed: CampaignSeed,
    agents: List<AgentProfile>,
    graph: SocialGraph,
    numTicks: Int = 3,
    generation: Int = 1,
): SimulationResult {
    val agentMap = agents.associateBy { it.id }
    val allInteractions = mutableListOf<Interaction>()
    val agentsWhoSawPost = mutableSetOf<String>()
    val influencerIds = graph.getInfluencers(5).map { it.id }.toSet()

    // Tick 0: initial exposure — influencers + random 30%
    val initialViewers = buildSet {
        addAll(influencerIds)
        addAll(agents.shuffled().take(agents.size / 3).map { it.id })
    }

    for (tick in 0 until numTicks) {
        val viewerIds: Set<String> = if (tick == 0) {
            initialViewers
        } else {
            // --- SECOND-ORDER SPREAD ---
            // Agents see the post if someone they follow shared or quoted it
            val sharers = allInteractions
                .filter { it.action in listOf(ActionType.SHARE, ActionType.QUOTE_SHARE) }
                .map { it.agentId }
                .toSet()

            val commenters = allInteractions
                .filter { it.action in listOf(ActionType.COMMENT, ActionType.MOCK) }
                .map { it.agentId }
                .toSet()

            val newViewers = mutableSetOf<String>()

            // Everyone following a sharer sees it (share = explicit distribution)
            for (sharerId in sharers) {
                val sharer = agentMap[sharerId] ?: continue
                newViewers.addAll(sharer.followers.filter { it !in agentsWhoSawPost })
            }

            // 30% chance the algorithm surfaces it to followers of commenters
            // (simulates algorithmic amplification)
            for (commenterId in commenters) {
                val commenter = agentMap[commenterId] ?: continue
                for (followerId in commenter.followers) {
                    if (followerId !in agentsWhoSawPost && Random.nextDouble() < 0.3) {
                        newViewers.add(followerId)
                    }
                }
            }

            newViewers
        }

        println("    Tick $tick: ${viewerIds.size} agents viewing")

        for (agentId in viewerIds) {
            if (agentId in agentsWhoSawPost && tick > 0) continue
            agentsWhoSawPost.add(agentId)

            val agent = agentMap[agentId] ?: continue

            // What can this agent see?
            // Reactions from people they follow + reactions from top influencers
            val visible = allInteractions.filter { ix ->
                ix.agentId in agent.following || ix.agentId in influencerIds
            }

            val context = Prompts.buildContext(seed, visible, agentMap)

            val (action, content, newMood) = simulateAgentReaction(executor, agent, context)
            agent.mood = newMood

            allInteractions.add(
                Interaction(
                    agentId = agent.id,
                    action = action,
                    content = content,
                    tick = tick,
                )
            )
        }
    }

    return SimulationResult(
        generation = generation,
        seed = seed,
        interactions = allInteractions,
    ).also { it.computeAnalytics() }
}

// ---------------------------------------------------------------------------
// Run a single agent's reaction through a Koog AIAgent
// ---------------------------------------------------------------------------

private suspend fun simulateAgentReaction(
    executor: LLMPromptExecutor,
    agent: AgentProfile,
    context: String,
): Triple<ActionType, String, Mood> {
    val systemPrompt = buildString {
        appendLine("You are roleplaying as ${agent.name}, a ${agent.age}-year-old from ${agent.location}.")
        appendLine("Bio: ${agent.bio}")
        appendLine("You respond ONLY with valid JSON. No markdown, no explanation.")
    }

    val userPrompt = Prompts.agentBrain(agent, context)

    val koogAgent = AIAgent(
        promptExecutor = executor,
        systemPrompt = systemPrompt,
        llmModel = OpenAIModels.Chat.GPT4o,
    )

    return try {
        val raw = koogAgent.run(userPrompt)
        parseAgentAction(raw, agent.mood)
    } catch (e: Exception) {
        println("      [WARN] Agent ${agent.name} errored: ${e.message}")
        Triple(ActionType.IGNORE, "", agent.mood)
    }
}

// ---------------------------------------------------------------------------
// JSON parsing
// ---------------------------------------------------------------------------

private val jsonParser = Json { ignoreUnknownKeys = true; isLenient = true }

fun parseAgentAction(raw: String, fallbackMood: Mood): Triple<ActionType, String, Mood> {
    val cleaned = raw
        .replace("```json", "")
        .replace("```", "")
        .trim()

    return try {
        val resp = jsonParser.decodeFromString<AgentActionResponse>(cleaned)
        Triple(
            ActionType.fromValue(resp.action),
            resp.content,
            Mood.fromValue(resp.new_mood),
        )
    } catch (e: Exception) {
        Triple(ActionType.IGNORE, "", fallbackMood)
    }
}
